### PR TITLE
Feature/add support for environment management

### PR DESCRIPTION
**Pull request by kentico-lubosr**
### Motivation

Fixes [#136](https://github.com/Kentico/kontent-management-sdk-net/issues/136)

### Checklist

- [ ] Code follows coding conventions held in this repo
- [ ] Automated tests have been added
- [ ] Tests are passing
- [ ] Docs have been updated (if applicable)
- [ ] Temporary settings (e.g. variables used during development and testing) have been reverted to defaults

### How to test

If manual testing is required, what are the steps?

### CODE REVIEW

File: Kentico.Kontent.Management/IManagementClient.cs [hunk 1]
```diff
@@ -557,5 +559,33 @@ public interface IManagementClient
...
+
+        /// <summary>
+        /// Clones environment.
+        /// </summary>
+        /// <param name="cloneEnvironmentModel">Cloning settings.</param>
+        Task<EnvironmentClonedModel> CloneEnvironmentAsync(EnvironmentCloneModel cloneEnvironmentModel);
+
+        /// <summary>
+        /// Returns the state of the environment cloning.
+        /// </summary>
```
Comment (gormal) [inline Kentico.Kontent.Management/IManagementClient.cs:571]: < returns> is missing in XML docs

File: Kentico.Kontent.Management/IManagementClient.cs [hunk 2]
```diff
@@ -557,5 +559,33 @@ public interface IManagementClient
...
+
+        /// <summary>
+        /// Deletes current environment.
+        /// </summary>
+        Task DeleteEnvironmentAsync();
+
+        /// <summary>
+        /// Marks current environment as production.
+        /// </summary>
+        Task MarkEnvironmentAsProductionAsync(MarkAsProductionModel markAsProductionModel);
```
Comment (gormal) [inline Kentico.Kontent.Management/IManagementClient.cs:582]: < param> is missing in XML docs

File: Kentico.Kontent.Management/Models/Environments/EnvironmentClonedModel.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        /// Gets or sets the ID of the cloned environment.
+        /// </summary>
+        [JsonProperty("id")]
+        public Guid Id { get; set; }
+
+        /// <summary>
+        /// Gets or sets the Management API key.
+        /// </summary>
+        [JsonProperty("management_api_key")]
+        public string ManagementApiKey { get; set; }
```
Comment (gormal) [inline Kentico.Kontent.Management/Models/Environments/EnvironmentClonedModel.cs:21]: Is this safe? As we only support HTTPS for MAPI I guess so...

**Review by gormal — CHANGES_REQUESTED** [review]
Thank you for your contribution! 

Changes look nice, I have only a few notes/questions. Don't hesitate to contact me if something is unclear.

File: Kentico.Kontent.Management/IManagementClient.cs [hunk 4]
```diff
@@ -557,5 +559,33 @@ public interface IManagementClient
         /// <param name="identifier">The identifier of the subscription user.</param>
         /// <returns></returns>
         Task DeactivateSubscriptionUserAsync(UserIdentifier identifier);
+
+        /// <summary>
+        /// Clones environment.
+        /// </summary>
+        /// <param name="cloneEnvironmentModel">Cloning settings.</param>
```
[the remark this answers is not in the document]
Comment (kentico-lubosr) [reply Kentico.Kontent.Management/IManagementClient.cs:566]: < returns> added

File: Kentico.Kontent.Management/IManagementClient.cs (hunk 1 above)
Comment (kentico-lubosr) [reply Kentico.Kontent.Management/IManagementClient.cs:572]: < returns> added

File: Kentico.Kontent.Management/IManagementClient.cs (hunk 2 above)
Comment (kentico-lubosr) [reply Kentico.Kontent.Management/IManagementClient.cs:585]: < param> added

File: Kentico.Kontent.Management/IManagementClient.cs [hunk 5]
```diff
@@ -557,5 +559,33 @@ public interface IManagementClient
...
+
+        /// <summary>
+        /// Marks current environment as production.
+        /// </summary>
+        Task MarkEnvironmentAsProductionAsync(MarkAsProductionModel markAsProductionModel);
+
+        /// <summary>
+        /// Modifies current environment.
+        /// </summary>
+        /// <param name="changes">Represents changes that will be applied to the environment.</param>
```
[the remark this answers is not in the document]
Comment (kentico-lubosr) [reply Kentico.Kontent.Management/IManagementClient.cs:590]: < returns> added

File: Kentico.Kontent.Management.Tests/Unit/ManagementClientTests/EnvironmentTests.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        private FileSystemFixture _fileSystemFixture;
+
+        public EnvironmentTests(FileSystemFixture fileSystemFixture)
+        {
+            _fileSystemFixture = fileSystemFixture;
+            _fileSystemFixture.SetSubFolder("Environment");
+        }
+
+        [Fact]
+        public async void CloneEnvironment_ReturnsNewEnvironment()
```
[the remark this answers is not in the document]
Comment (kentico-lubosr) [reply Kentico.Kontent.Management.Tests/Unit/ManagementClientTests/EnvironmentTests.cs:30]: Test added

File: Kentico.Kontent.Management.Tests/Unit/ManagementClientTests/EnvironmentTests.cs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            var client = _fileSystemFixture.CreateMockClientWithoutResponse();
+
+            Func<Task> deleteEnvironment = async () => await client.DeleteEnvironmentAsync();
+
+            await deleteEnvironment.Should().NotThrowAsync();
+        }
+
+        [Fact]
+        public async void ModifyEnvironment_Rename_RenamesEnvironment()
```
[the remark this answers is not in the document]
Comment (kentico-lubosr) [reply Kentico.Kontent.Management.Tests/Unit/ManagementClientTests/EnvironmentTests.cs:96]: Test added

File: Kentico.Kontent.Management/ManagementClient.cs [hunk 8]
```diff
@@ -1072,6 +1074,51 @@ public async Task DeactivateSubscriptionUserAsync(UserIdentifier identifier)
...
+
+        /// <inheritdoc />
+        public async Task DeleteEnvironmentAsync()
+        {
+            var endpointUrl = _urlBuilder.BuildProjectUrl();
+            await _actionInvoker.InvokeMethodAsync(endpointUrl, HttpMethod.Delete);
+        }
+
+        /// <inheritdoc />
+        public async Task MarkEnvironmentAsProductionAsync(MarkAsProductionModel markAsProductionModel)
```
[the remark this answers is not in the document]
Comment (kentico-lubosr) [reply Kentico.Kontent.Management/ManagementClient.cs:1104]: Yes, I've added the check and a test too.

File: Kentico.Kontent.Management/Models/Environments/EnvironmentClonedModel.cs (hunk 3 above)
Comment (kentico-lubosr) [reply Kentico.Kontent.Management/Models/Environments/EnvironmentClonedModel.cs:21]: We consulted this with the security team when we designed the endpoint and they consider that safe.

File: Kentico.Kontent.Management/Models/Environments/EnvironmentModel.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        public Guid Id { get; set; }
+
+        /// <summary>
+        /// Gets or sets the name of the environment
+        /// </summary>
+        [JsonProperty("name")]
+        public string Name { get; set; }
+
+        /// <summary>
+        /// Gets or sets Gets or sets a flag determining whether the environment is production
```
[the remark this answers is not in the document]
Comment (kentico-lubosr) [reply Kentico.Kontent.Management/Models/Environments/EnvironmentModel.cs]: Fixed

File: Kentico.Kontent.Management/Models/Environments/MarkAsProductionModel.cs [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+{
+    /// <summary>
+    /// Represents settings of marking environment as production.
+    /// </summary>
+    public class MarkAsProductionModel
+    {
+        /// <summary>
+        /// Gets or sets a flag determining whether webhooks on the new production environment should be enabled.
+        /// </summary>
+        [JsonProperty("enable_webhooks", Required = Required.Always)]
```
[the remark this answers is not in the document]
Comment (kentico-lubosr) [reply Kentico.Kontent.Management/Models/Environments/MarkAsProductionModel.cs]: Required flag removed

**Review by gormal — APPROVED** [review]

**Pull request merged**